content/post/2020-11-18-testthat-utility-belt/index.Rmd

```diff
@@ -0,0 +1,106 @@
+
+To summarize,
+
+| File                           | Run before tests | Loaded via `load_all()` | Installed with the package[^installed] | Testable[^testable] |
+|--------------------------------|------------------|-------------------------|----------------------------|----------|
+| tests/testthat/setup-.\*.R     | ✔️               | \-                      | \-                         | \-       |
+| tests/testthat/helper-.\*.R    | ✔️               | ✔️                      | \-                         | \-       |
+| R/any-name.R                   | ✔️               | ✔️                      | ✔️                         | ✔️       |
+| tests/testthat/anything-else.R | \-               | \-                      | \-                         | \-       |
+
+`tests/testthat/helper-.*.R` are [no longer recommended in testthat](https://testthat.r-lib.org/reference/test_dir.html#special-files) but they are still supported. :relieved:
+
+[^installed]: Installed with the package, and [run at package installation time](https://github.com/r-lib/testthat/issues/1206#issuecomment-713519962).
+[^testable]: Yes, you could [test the code supporting your tests](https://github.com/r-lib/testthat/issues/1206#issuecomment-713583205).
+
+In practice,
+
+* In `tests/testthat/setup.R` you might do something like loading a package that helps your unit testing like `{vcr}`, `{httptest}` or `{presser}` if you're [testing an API client](https://books.ropensci.org/http-testing/).
+* In a helper like `tests/testthat/helper.R` or `R/test-helpers.R` you might define variables and functions that you'll use throughout your tests, even [custom skippers](https://testthat.r-lib.org/articles/skipping.html#helpers). To choose between the two locations, refer to the table above and your own needs and preferences. Note that if someone wanted to study testthat "utility belts" à la [Bob Rudis](https://rud.is/b/2018/04/08/dissecting-r-package-utility-belts/), they would probably only identify helper files like `tests/testthat/helper.R`.
+
+You'll notice testthat no longer recommends having a file with code to be run after tests... So how do you clean up after tests?
+Well, with `withr`'s various helper functions for deferring clean-up.
+So basically it means the code for cleaning lives near the code for making a mess.
+To learn more about this, read [the "self-cleaning text fixtures" vignette in testthat](https://testthat.r-lib.org/articles/test-fixtures.html).
```

Comment:
Should we have an example and also mention `teardown_env()`?

Reply:
I think the vignette is the best resource so won't add any example.